Improve automatic local referencing

vinaysraghavan opened the pull request:
<!--
Thanks for contributing a pull request to naplib-python!
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.
Change the make_contact_rereference_arr function to handle input channelnames in any order (not strictly ascending) and to properly handle local referencing on ECoG grid

#### Any other comments?

vinaysraghavan committed:
Update rereference.py

Change the make_contact_rereference_arr function to handle input channelnames in any order (not strictly ascending) and to properly handle local referencing on ECoG grid

vinaysraghavan committed:
Update rereference.py

Added input `grid_sizes` to allow custom input of grid sizes. After talking with Max Nentwich, I realized that, while square and 1 x 2 rectangle grids are relatively standard/common, grids are occasionally cut, so custom shapes as input is needed.

vinaysraghavan committed:
Update test_rereference.py

Update test of make_contact_rereference_arr to improve coverage

vinaysraghavan committed:
Update test_rereference.py

Change expected test outcome

vinaysraghavan committed:
Updated rerefererence array test to improve coverage

vinaysraghavan committed:
Improve test coverage for make contact rereference array

vinaysraghavan committed:
Improve test coverage further

vinaysraghavan committed:
Remove extra lines

gavinmischler commented on naplib/preprocessing/rereference.py:
```diff
@@ -145,18 +148,80 @@ def make_contact_rereference_arr(channelnames, extent=None):
+                    adjacent = _find_adjacent_numbers(nrows, ncols, ch, extent)
+                    curr = np.where(channelnames==f'{contact}{ch}')[0]
+                    inds = []
+                    print(ch, adjacent)
```
Is this print statement from debugging? Can it be removed?

vinaysraghavan replied:
Oh yes, good catch

gavinmischler commented on tests/preprocessing/test_rereference.py:
```diff
@@ -4,10 +4,29 @@
 
 
 def test_create_contact_rereference_arr():
```
can you create a 2nd test for this and keep the old test, just to confirm old behavior where no ``grid_sizes`` was provided still works?

vinaysraghavan replied:
It's included in the same test function. The `expected1` and `arr1` variables. I changed the "RT"s to "GridA"s just to get more test coverage when including grid_sizes, but the test for the original behavior is still there

gavinmischler replied:
Oh I see, thanks

gavinmischler reviewed (commented):
Just a couple small things, then looks good!

vinaysraghavan committed:
Update rereference.py

Remove debugging print statement

gavinmischler reviewed (approved):
Looks good!

vinaysraghavan committed:
Update rereference.py

Removed the assumption that channels 1-N are always provided for a given contact. Now, any channel numbers on a contact can be provided, and local average referencing will be robust to this. 

This is useful in cases where depth electrodes are inserted through the brain, e.g. some electrodes are in the CSF beyond the brain, some electrodes are in the brain, and some are in the skull/scalp, since only electrodes in the brain should be used for re-referencing.

It now also notifies the user if channels cannot properly be re-referenced because `extent` is too small due to a lack of spatially local sites.
